staging-service/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java

```diff
@@ -0,0 +1,92 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.cafdataprocessing.services.staging.utils;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.commons.fileupload.ProgressListener;
+import org.apache.commons.fileupload.servlet.ServletFileUpload;
+
+public final class BatchProgressTracker
+{
+    private static final ConcurrentHashMap<String, Tracker> inProgressTrackerMap = new ConcurrentHashMap<>();
+
+    public void updateTracker(final ServletFileUpload fileUpload)
+    {
+        fileUpload.setProgressListener(new ProgressListenerImpl());
+    }
+
+    public void remove(final long threadId, final String serviceId)
+    {
+        inProgressTrackerMap.remove(threadId + "-" + serviceId);
+    }
+
+    public void putAll(final Map<String, Tracker> tempMap)
+    {
+        inProgressTrackerMap.putAll(tempMap);
+    }
+
+    public void removeTrackerOfDifferentService()
+    {
+        inProgressTrackerMap.entrySet().removeIf(entry -> !entry.getKey().contains(ServiceIdentifier.getServiceId()));
+    }
+
+    public Tracker get(final String batchThread)
+    {
+        return inProgressTrackerMap.get(batchThread);
+    }
+
+    public void put(final String threadID, final Tracker tracker)
+    {
+        inProgressTrackerMap.put(threadID, tracker);
+    }
+
+    private static class ProgressListenerImpl implements ProgressListener
+    {
+        private final Tracker tracker;
+        private long megaBytes;
+
+        public ProgressListenerImpl()
+        {
+            final Tracker newTracker = new Tracker();
+            newTracker.setUploadStartTime(Instant.now());
+            this.tracker = newTracker;
+            this.megaBytes = -1;
+        }
+
+        @Override
+        public void update(final long totalBytesRead, final long contentLength, final int item)
+        {
+            final long totalBytesInMBytes = totalBytesRead / 1_048_576;
+            if (megaBytes == totalBytesInMBytes) {
+                return;
+            }
+            megaBytes = totalBytesInMBytes;
+            final Instant now = Instant.now();
+            final long diff = ChronoUnit.MILLIS.between(tracker.getUploadStartTime(), now);
+            if (diff > 0) {
+                final long speed = Math.floorDiv(totalBytesRead, diff);
+                tracker.setFileUploadRateInBytesPerSecond(speed * 1000);
+                tracker.setNumberOfBytesReceived(totalBytesRead);
+                tracker.setLastModifiedTime(now);
+                tracker.setProgressing(true);
```

Comment:
Unless this is the first `update()` call, you are updating a `Tracker` object that is already in the `ConcurrentHashMap`.  But you are doing it piecemeal, one property at a time.  Another thread could be reading this information simultaneously, and therefore it could end up with an erroneous half-half report.  To do what you're trying to do you need to synchronize updates and reads in the `Tracker` object.  It would probably be better and easier to change the object to be immutable and to replace it in the map.

Reply:
**Resolved in this commit:**
[590f331c0db34ae520d5a2d8a635c62a07f8bd09] Code review: Extensive refactoring

The `Tracker` object has been updated to be immutable; all of its fields are now `final`.  The objects are now replaced in the map rather than being updated.